### ISSUE TITLE
LHN - Display chats with an existing draft under pinned chats in LHN 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#4615]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to a conversation that's not on the top of LHN 
2. Write something in the compose box. 
3. Navigate away from the chat 

## Expected Result:
Conversation with a draft message should be displayed on the top of LHN for visibility

## Actual Result:
Conversation is not displayed on the top and can be easily lost in the LHN conversation list. 

## Workaround:
User can manually look for the conversation with the drafted message. 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.0.85-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

![image](https://user-images.githubusercontent.com/44479856/129252212-5f462054-fada-43b3-9218-a98dbae1808c.png)

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @rafecolton https://expensify.slack.com/archives/C01GTK53T8Q/p1628786564265600
> Chats with draft messages no longer show at the top of the LHN below pinned chats. Unless I missed a conversation somewhere, this is a regression. Edit: I'm on macOS desktop 1.0.82-7

**Comment by MelvinBot** [comment]
Triggered auto assignment to @aldo-expensify (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by aldo-expensify** [comment]
I would think that this can be done by an external (upwork), but I would like to do it myself to learn more about the App project.

**Comment by aldo-expensify** [comment]
By looking at the code in https://github.com/Expensify/App/blob/ff62078baf05801bce5a5c3ebffb20f89b25bf0d/src/libs/OptionsListUtils.js#L387-L428

I can see that we have to choose one option to prioritize: `iou debt` or `pinned`, and now we are adding `with draft comment`. The way it is implemented now, looks like the prioritized is exclusive, if I prioritize by `pinned`, then 'iou debt' won't be prioritized.

How do we want the reports `with draft comment` to appear? if it is prioritized by 'iou debt', then the chat with draft comments should be right below the 'iou debt'? or on top?

Wouldn't it be better to be able to have also a way of keeping things in a prioritization like:

1. pinned
2. iou debt
3. drafts
4. all the rest

And then the user, in the settings, can check or uncheck if they want to have categories like `iou debt` prioritized or not. If a user uncheck  `iou debt` as a priority, then these reports get mixed git 'all the rest', leaving the reports in the following order:

1. pinned
2. drafts
3. all the rest

**Comment by MelvinBot** [comment]
Triggered auto assignment to @JmillsExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by aldo-expensify** [comment]
In my opinion, this can be performed by an external because it seems to be only UI stuff

**Comment by botify** [comment]
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1183084866) to staging by @iwiznia in version: 1.0.88-3 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|cancelled 🔪 
🖥 desktop 🖥|cancelled 🔪
🍎 iOS 🍎|cancelled 🔪 
🕸 web 🕸|success ✅

**Comment by botify** [comment]
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1188395063) to production by @roryabraham in version: 1.0.90-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

**Comment by mallenexpensify** [comment]
@meetmangukiya, on Desktop Version 1.0.92-0, this is what I'm experiencing. 
- Navigate to new chat that's not currently at the top of Left Hand Nav (LHN) 
- Start typing a message
- Once paused for 1-2 seconds, the new draft chat populates at the top of LHN. 

I'm assuming this is performing 'as expected', can you confirm?

Personally... I find it to be a bit distracting that, when I'm typing a message, I see a 'flash' in the LHN and new chat loaded at top, I often/always look over at the chat thinking I have a new one coming it, but it's just my current draft that I'm actively working on. Are there other options you considered or that you know are available for how to manage this?  

Can we do something like: 
1. When someone is composing a text and there are characters in the compose box, 'do nothing'.  
2. If a users leaves a chat and there are characters still in the box (aka a draft message), **then** move it up to the top of LHN.  That way it wouldn't 'flash' and distract a user when they're in the middle of composing a message.  I don't think there's much value in seeing a draft message while you're composing, only once you're navigated away from the chat . 

What do you think @meetmangukiya ? 
@aldo-expensify , @iwiznia, @mountiny 👀 too please, in case I'm missing something on the engineering/code side

**Comment by MelvinBot** [comment]
@iwiznia, @meetmangukiya, @mallenexpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by mallenexpensify** [comment]
Waiting for feedback from my previous comment

**Comment by mountiny** [comment]
@mallenexpensify Yes, that is expected behaviour I would say.

Honestly, I am not totally sure what would be the best way to approach this. Let's say we would reorder the LHN on choosing some other chat. If it would be chat which is located above the draft chat, we would see shift at time of clicking to the new message (because the one we clicked on would be shifted one below). 

I think this would be the best to ask in #expensify-open-source to get some thoughts and ideally ideas on how this is treated in other apps.

I have tried Facebook Messenger and Whatsapp desktop apps and they dont even do drafts. In Signal desktop App, it is the way as I described above, they reorder the LHN once user clicks on other chat causing it shift at that moment. Maybe that is a good compromise.

**Comment by meetmangukiya** [comment]
If we were to to want to implement this I suppose we could add a `shouldComponentUpdate` that will do some smart diffing in previous and new props to see if any new drafts, and if they are for current report, etc. and respectively return.

**Comment by mallenexpensify** [comment]
Posted in Slack, let's see what others think 
https://expensify.slack.com/archives/C01GTK53T8Q/p1631062473055500

**Comment by mallenexpensify** [comment]
@mountiny and I agree we should update this to:
1. When someone is composing a text and there are characters in the compose box, 'do nothing'.
2. If a users leaves a chat and there are characters still in the box (aka a draft message), then move it up to the top of LHN. That way it wouldn't 'flash' and distract a user when they're in the middle of composing a message. I don't think there's much value in seeing a draft message while you're composing, only once

@meetmangukiya are you able to do that? If so, do you want to do as part of this GH issue or would a new GH issue be better? Either way, I think you should be compensated for the update since it wasn't clearly stated that was what we originally were asking for and it will require additional work.  

Thanks @kidroca for the :yes: on the [Slack post](https://expensify.slack.com/archives/C01GTK53T8Q/p1631062473055500) too.

**Comment by meetmangukiya** [comment]
@mallenexpensify yes, I can do it. I am fine with either continuing this issue or creating a new one. Is it possible for the part 1(?) payment to be made while we implement, review and test the update?

**Comment by mallenexpensify** [comment]
@meetmangukiya 
1. Do you want to work on the update?
2. If so, do you have a preference between working off this existing issue or creating a brand new one? The pay will be the same, $250. 

@vitHoracek it's easy to add a bonus to a job and/or create a new job in Upwork

**Comment by mallenexpensify** [comment]
Paid @meetmangukiya for the original job in Upwork. Job is still open in Upwork til we decide next best step

**Comment by mallenexpensify** [comment]
ooooof, **completely** missed this comment above
> @mallenexpensify yes, I can do it. I am fine with either continuing this issue or creating a new one. Is it possible for the part 1(?) payment to be made while we implement, review and test the update?

I created a new issue (which is almost always 'best practice') https://github.com/Expensify/App/issues/5166
It will go through our normal process, I added a note that you have 'first dibs' to fix the issue @meetmangukiya so other contributors shouldn't propose fixes.  
Closing this now and also the upwork job (there will be a new upwork job for the other with separate payment)

**Comment by mountiny** [comment]
Thank you for handling and clarifying the process @mallenexpensify 🙌

**Comment by meetmangukiya** [comment]
Thank you!
